lib/api/src/modules/stories.ts

```diff
@@ -268,105 +322,63 @@ export const init: ModuleFn = ({
     // Later when we change story via the manager (or SELECT_STORY below), we'll already be at the
     // correct path before CURRENT_STORY_WAS_SET is emitted, so this is less important (the navigate is a no-op)
     // Note this is the case for refs also.
-    fullAPI.on(CURRENT_STORY_WAS_SET, function handleCurrentStoryWasSet({ storyId, viewMode }) {
-      const { source }: { source: string } = this;
-      const [sourceType] = getSourceType(source);
-
-      if (sourceType === 'local' && storyId && viewMode) {
-        navigate(`/${viewMode}/${storyId}`);
-      }
-    });
-
-    fullAPI.on(STORY_CHANGED, function handleStoryChange(storyId: string) {
-      const { source }: { source: string } = this;
-      const [sourceType] = getSourceType(source);
-
-      if (sourceType === 'local') {
-        const options = fullAPI.getCurrentParameter('options');
-
-        if (options) {
-          fullAPI.setOptions(options);
+    fullAPI.on(
+      CURRENT_STORY_WAS_SET,
+      createEventHandlerWithRefSupport<{ storyId: string; viewMode: ViewMode; [k: string]: any }>(
+        ({ storyId, viewMode }, { sourceType }) => {
```

Comment:
This feels like an unnecessary level of indirection/complexity. Couldn't we just do this:
```suggestion
    fullAPI.on(CURRENT_STORY_WAS_SET, ({ storyId, viewMode }) => {
      const { sourceType } = getRefFromEvent(this);
```

Reply:
`this` can only be used in non-arrow-functions; Sure, I can make that change everywhere.

Reply:
Oh, yep, but you get the idea.

Reply:
I do 👍

Reply:
done, WDYT?